community/graphdb-api/src/main/java/org/neo4j/graphdb/PropertyContainer.java

```
@@ -145,4 +145,15 @@
      * @return all properties on this property container
      */
     Map<String, Object> getAllProperties();
+
+    /**
+     * Returns the unique id of this <code>PropertyContainer</code>. Ids are garbage collected over time
+     * so they are only guaranteed to be unique during a specific time span: if
+     * the  <code>PropertyContainer</code> is deleted, it's likely that a new
+     * <code>PropertyContainer</code> at some point will get the old id. <b>Note</b>: this makes
+     * <code>PropertyContainer</code> ids brittle as public APIs.
+     *
+     * @return The id of this  <code>PropertyContainer</code>.
+     */
+    long getId();
```

You can't actually just define a new method on an interface like this, since it breaks compatibility with implementors. In fact in Neo4j itself there are more implementors of `PropertyContainer` than just `Node` and `Relationship` sub-types.

It needs to be a `default` method on the interface, and I think it should throw `UnsupportedOperationException` in its default implementation. This of course also needs to be documented; that it may throw UOE if the entity in question does not have an Id.

---

Instead of adding a `default` implementation throwing `UnsupportedOperationException`, maybe we can add an extra interface that extends `PropertyContainer` and has the `getId` method. Only `Node`/`Relationship` (and similar) will implement this new interface and we leave untouched the other implementation of `PropertyContainer`.

---

While both suggestions are superior to what I proposed, davidegrohmann's accomplishes the same goal while being the least invasive. I will revert my original changes and implement davidegrohmann's suggestion for review.

---

The proposed changes have been coded and pushed.

---

You didn't push your changes to the branch that you issued this pull request from, so they are not showing up in this PR. Also, I think `IdentifiablePropertyContainer` would be a better name.

---

I have changed the name to IdentifiablePropertyContainer, committed, and pushed to the PR branch.